libsplinter/src/error/constraint_violation.rs

```diff
@@ -0,0 +1,202 @@
+/// a table with a unique column, and an attempt to insert an entry which would cause duplication in
+/// that column, an error with violation type `ConstraintViolationType::Unique` will occur.
+///
+/// Although this error maps closely to those generated by relational databases (such as those
+/// covered by Diesel), the underlying database does not need to be relational. It could, for
+/// example, be a memory or file-backed implementation of a store.
+pub struct ConstraintViolationError {
+    violation_type: ConstraintViolationType,
+    source: Option<Box<dyn error::Error>>,
+}
+
+impl ConstraintViolationError {
+    /// Constructs a new `ConstraintViolationError` from a specified violation type.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will use the
+    /// standard display of the ConstraintViolationType for its message.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use splinter::error::{ ConstraintViolationError, ConstraintViolationType };
+    ///
+    /// let constraint_violation_error = ConstraintViolationError::with_violation_type(
+    ///     ConstraintViolationType::Unique
+    /// );
+    /// assert_eq!(format!("{}", constraint_violation_error), "Unique constraint violated");
+    /// ```
+    pub fn with_violation_type(violation_type: ConstraintViolationType) -> Self {
+        Self {
+            violation_type,
+            source: None,
+        }
+    }
+
+    /// Constructs a new `ConstraintViolationError` from a specified source error and violation type.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will simply pass through the
+    /// display of the source message unmodified.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use splinter::error::{ ConstraintViolationError, ConstraintViolationType };
+    ///
+    /// let db_err = std::io::Error::new(std::io::ErrorKind::Other, "db error");
+    /// let constraint_violation_error = ConstraintViolationError::from_source_with_violation_type(
+    ///     ConstraintViolationType::Unique,
+    ///     Box::new(db_err)
+    /// );
+    /// assert_eq!(format!("{}", constraint_violation_error), "db error");
+    /// ```
+    pub fn from_source_with_violation_type(
+        violation_type: ConstraintViolationType,
+        source: Box<dyn error::Error>,
+    ) -> Self {
+        Self {
+            violation_type,
+            source: Some(source),
+        }
+    }
+}
+
+impl error::Error for ConstraintViolationError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match &self.source {
+            Some(s) => Some(s.as_ref()),
+            None => None,
+        }
+    }
+}
+
+impl fmt::Display for ConstraintViolationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.source {
+            Some(s) => write!(f, "{}", s),
+            None => write!(f, "{} constraint violated", &self.violation_type),
```

Comment:
"Data Type: xxxx constraint violated" is awkward

Reply:
Should I just get rid of the DataType enum?